### ISSUE TITLE
Abnormal Memory Demand  

### ISSUE THREAD

**Issue by ben0i0d** [alonfnt/tsnex#7]
When using t-SNE on a dataset of shape (2500, 784) on the CPU, I found that the memory usage peaked at around 34.5GB, which made it impossible to run the program on my local computer.  

This is quite unreasonable since the dataset is small. Furthermore, I did not observe such behavior with sklearn or pcax; they showed almost no memory demand.  

I have uploaded my code and process to [https://github.com/ben0i0d/tsne-test](https://github.com/ben0i0d/tsne-test).  

If you have time, please take a look.

**Comment by ben0i0d** [comment]
I am using

JAX version 0.4.37
pcax  version 0.1.0
tsnex version 0.0.1

installed via pip.

**Comment by ben0i0d** [comment]
related to

```
X_new, _ = jax.lax.scan(train_step_early_exageration, X_new, xs=None, length=n_exageration)  
X_new, _ = jax.lax.scan(train_step, X_new, xs=None, length=n_iter)  
```

**Comment by alonfnt** [comment]
Hi @ben0i0d 
Thanks for raising this and thanks a lot for the repo to test it.

I agree with you that it is unresonable, but sadly it is working "as expected". 
The current implementation (which is quite simple since I needed it to compute some gradients), relies on computing the distance matrix directly. You can get the same error if you run 

```python
dist_fn = jax.vmap(jax.vmap(tsnex.tsne.euclidean_distance, in_axes=(0, None)), in_axes=(None, 0))
D = dist_fn(x,x)
```

with some napkin calculations: assuming single floating point (32bits), to compute D we need to create a (2500, 2500, 784) matrix, which should give ~20Gb, then we add D and we get the >30Gb. 

The reason sklearn can get away with it is because, other than all the approximations method that I need to include #4 #5 , they rely on scipy distance calculations (when using `method='exact'` as an argument, which would be equivalnet to tsnex), and scipy has a neat optimization for square matrices distances. 


Now, what can we do... The first simple solution could be to check the dimension of x, and if above a threshold k, split D to compute every k rows (that would save some memory).

Ideally, I think more optimize distance calculations should be implemented, but that takes time.

If you feel to contribute I'd gladly review the above "quick" solution, otherwise I'll give it a go soon.

Thanks for pointing this out and let's try to solve it!

**Comment by ben0i0d** [comment]
Thank you for your reply. I also noticed related to the distance computation:
`metric_fn = jax.vmap(jax.vmap(metric_fn, in_axes=(0, None)), in_axes=(None, 0))`

If I have time later, I would be interested in learning more about the optimization methods you mentioned （#4 #5） and trying some code.

Since I am working on an ROCm platform and almost all tSNE acceleration implementations are written for CUDA, I also tried the tSNE-Torch implementation. Unfortunately, it took 2 minutes and 13 seconds on this dataset, whereas this JAX implementation is extremely fast, taking only 0.1 seconds (so I truly appreciate your work).

**Comment by ben0i0d** [comment]
Although I am still busy with work, I managed to solve this issue today. Let me briefly introduce it.

In fact, we don't need to use vmap for overall vectorization. According to the description of jax.lax.map in the documentation, we can achieve this step by using the batch_size to ensure that no excessively large intermediate memory allocations occur.

Here’s a screenshot of my code.

![Image](https://github.com/user-attachments/assets/cda5cc40-68f8-40b2-910b-6c3b853d9961)

`jax.lax.map(lambda xi: jax.lax.map(lambda xj: euclidean_distance(xi, xj), x, batch_size = 512), x, batch_size = 512)`

Below are the results from my memory profiling:

vmap would cause about 10GB of intermediate memory allocation, while jax.lax.map with a batch_size of 512 only requires about 200MB.

Of course, I admit that this might be slightly slower. In theory, the batch_size size should be estimated based on the available memory to accelerate the process.

By using this approach, the code can run steadily without worrying about memory overflow. I hope this is helpful to you.

**Comment by alonfnt** [comment]
Hi @ben0i0d,
That's a good solution for now! I'd be up to include a `batch_size: Optional[int] = None` parameter to `tsnex.transform()` and replace the current line with something like
```python
batch_size = batch_size or len(X)
metric_fn = jax.lax.map(
    lambda xi: jax.lax.map(lambda xj: euclidean_distance(xi, xj), x, batch_size),
    x,
    batch_size,
)
```

that way we don't break current workflow in the meantime.

I think you deserve the contribution though so feel free to make a PR and I'll gladly review it :) -- I think it is time for a new release too.
